[Merged by Bors] - feat(algebra/lie/basic): define the center of a Lie algebra and prove some related results

ocfnash opened the pull request:
---
I had originally intended to include the definition of reductive Lie algebras. Ultimately I decided against this because I learned that:

- There is absolutely no agreement on [what the label "reductive" should mean](https://mathoverflow.net/questions/284713/is-there-a-reasonable-way-to-define-reductive-lie-algebra-in-prime-characteris) outside of fields of characteristic 0.
- The various conditions that are equivalent for a field of char 0 are not equivalent in general.
- It would require a moderate amount of work to establish equivalence of the conditions when sufficient assumptions on the coefficients are provided. (Even basic facts like the fact that radical is a characteristic ideal need some work.)

So I think this is a reasonable stopping point before I attempt [the refactor I mentioned](https://github.com/leanprover-community/mathlib/pull/5794#discussion_r560050825).

ocfnash committed:
feat(algebra/lie/basic): define the center of a Lie algebra and prove some related results

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -2013,6 +2091,10 @@ begin
 end
 
+lemma semisimple_lie_abelian_trivial [is_semisimple R L] [h : is_lie_abelian L] :
+  (⊤ : lie_ideal R L) = ⊥ :=
```
Is this the best way to express the fact that a Lie algebra is the zero Lie algebra? Should I have a separate predicate for this?

jcommelin replied:
I think we can use `subsingleton` for this. In principle the lemma could then become an `instance`, but we recently discovered that too many subsingleton instances slows down the library. So maybe it should stay a `lemma` for now.

ocfnash replied:
OK I'll leave it as it. I suppose the `subsingleton` approach would be to prove:
```lean
instance [is_semisimple R L] [is_lie_abelian L] : subsingleton (lie_algebra R L) :=
```
I think I can imagine ways in which this would make the typeclass inference engine's job a lot harder.

eric-wieser replied:
I think @jcommelin is suggesting you use `subsingleton` without `instance`.

ocfnash replied:
Oh I see. I'll add that now.

Update: this is not quite totally trivial, I'll have to return to it later today when I have more time.

ocfnash replied:
On reflection it is not clear to me I should prove a subsingleton lemma. Indeed IIUC, the statement that I suggested above:
```lean
lemma [is_semisimple R L] [is_lie_abelian L] : subsingleton (lie_algebra R L) :=
```
only demonstrates that two semisimple Abelian Lie algebras _with the same underlying Lie ring structures_ have the same module structure.

I guess I could also prove:
```lean
lemma [is_semisimple R L] [is_lie_abelian L] : subsingleton (lie_ring R L) :=
```
but I begin to wonder what all this does for me.

I'm happy to do it of course if you think I should.

eric-wieser replied:
> Update: this is not quite totally trivial

can't you use `subsingleton_of_bot_eq_top` on your existing proof?

eric-wieser replied:
The statement is
```lean
lemma semisimple_lie_abelian_trivial [is_semisimple R L] [h : is_lie_abelian L] : subsingleton (lie_ideal R L) :=
subsingleton_of_bot_eq_top (your proof)
```

ocfnash replied:
Oh I see; mathematically that's weaker than what I had in mind but will be far more useful. Thanks!

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -2021,6 +2103,19 @@ begin
+by { rw is_semisimple.semisimple, exact is_lie_abelian_bot R L, }
+
+lemma abelian_radical_iff_solvable_is_abelian [is_noetherian R L] :
+  is_lie_abelian (radical R L) ↔ ∀ (I : lie_ideal R L), is_solvable R I → is_lie_abelian I :=
```
This is the closest I get to saying something about reductive Lie algebras.

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -1822,6 +1871,29 @@ def top_equiv_self : (⊤ : lie_ideal R L) ≃ₗ⁅R⁆ L :=
+/-- The center of a Lie algebra is the set of elements that commute with everything. It can
+be viewed as the maximal trivial submodule of the Lie algebra as a Lie module over itself via the
+adjoint representation. -/
+abbreviation center : lie_ideal R L := lie_module.maximal_trivial_submodule R L L
```
I have caved in and gone for the US spelling 😢

eric-wieser commented on src/algebra/lie/basic.lean:
```diff
@@ -562,6 +562,9 @@ def lie_module.to_endo_morphism : L →ₗ⁅R⁆ module.End R M :=
   map_smul' := λ t x, by { ext m, apply smul_lie, },
   map_lie   := λ x y, by { ext m, apply lie_lie, }, }
 
+@[simp] lemma lie_module.to_endo_morphism_apply (x : L) (m : M) :
```
I think you can add `@[simps]` on the def above to get this automatically - but maybe it picks the wrong name.

ocfnash replied:
Thanks; I had forgotten about `simps`! Trying now, I see it names the lemma `lie_module.to_endo_morphism_to_fun_apply` which is fine by me.

eric-wieser replied:
You can train simps to use `apply` instead of `to_fun` there - have a look at how we do it for equiv and linear maps.

ocfnash replied:
Oh I see we even have a [library note](https://leanprover-community.github.io/mathlib_docs/notes.html#custom%20simps%20projection) on this.

I just tried:
```lean
initialize_simps_projections lie_module.to_endo_morphism (to_fun_apply → apply)
```
and it just says `failed`.

Is it really worth going to trouble to persuade `simps` to name things differently in this case? I'm happy with `_to_fun_apply` and if I really want to control the name, I'd rather just have a one-liner defining a `rfl` lemma than grappling with meta code.

eric-wieser replied:
It should be something more like
```
initialize_simps_projections lie_algebra.morphism (to_fun → apply)
```
(I don't remember what `→ₗ⁅R⁆` is called, I assume its `lie_algebra.morphism`)

ocfnash replied:
Thanks, I finally get the idea. Now that I understand this better, I would like to use `initialize_simps_projections` except that it slightly changes the type of lemma that is generated by having different coercions in place❗️

Specifically, without `initialize_simps_projections`, I get a lemma named `lie_module.to_endo_morphism_to_fun_apply` which states:
```lean
⇑(⇑(lie_module.to_endo_morphism R L M) x) m = ⁅x,m⁆
```
However, with `initialize_simps_projections`, I get a lemma named `lie_module.to_endo_morphism_apply_apply` which states:
```lean
⇑((lie_module.to_endo_morphism R L M).to_fun x) m = ⁅x,m⁆
```

The first lemma (with the coercion instead of `to_fun`) is preferable; indeed in the proof of `lie_module.mem_ker` I have to change `simp` to `simpa` to close the goal if I use `lie_module.to_endo_morphism_apply_apply`.

eric-wieser replied:
That shouldn't be happening - I think you can fix it by using `initialize_simps_projections` _after_ creating the `has_coe_to_fun` instance. It's hard to debug without seeing your code though.

ocfnash replied:
I can confirm that invoking `initialize_simps_projections` after creating the `has_coe_to_fun` resolves this issue, so I've pushed a commit which does this now. This is exactly the sort of meta code issue that makes me wary but perhaps I'm just too paranoid.

eric-wieser replied:
I agree it's a little fragile, but we already use `initialize_simps_projections` on all the other bundled homs, so it's worth putting up with the quirks for consistency with `add_monoid_hom`, `linear_map`, `equiv`, ...

ocfnash replied:
Definitely in favour of consistency!

eric-wieser commented on src/linear_algebra/basic.lean:
```diff
@@ -161,6 +161,13 @@ instance unique_of_left [subsingleton M] : unique (M →ₗ[R] M₂) :=
 coe_injective.unique
 
+lemma eq_zero_iff (f : M →ₗ[R] M₂) : f = 0 ↔ ∀ m, f m = 0 :=
+begin
```
I think `linear_map.ext_iff` might be a proof of this lemma

ocfnash replied:
Basically yes.

In fact a proof is `simp only [ext_iff, zero_apply]` which convinces me that this lemma shouldn't even exist so I'll drop it.

ocfnash committed:
Use `simps` on `lie_module.to_endo_morphism` to avoid manual `_apply` lemma.

ocfnash committed:
Do not add `linear_map.eq_zero_iff`.

On reflection, I don't believe we need this lemma given that a proof is
`simp only [ext_iff, zero_apply]`.

ocfnash committed:
CR suggestions: use `subsingleton` and custom `simps` projections.

eric-wieser commented on src/algebra/lie/basic.lean:
```diff
@@ -983,16 +989,13 @@ by { rw [← mem_coe_submodule, sup_coe_to_submodule, submodule.mem_sup], exact
 by { rw eq_bot_iff, exact iff.rfl, }
 
-lemma of_bot_eq_bot (N : lie_submodule R L ↥(⊥ : lie_submodule R L M)) : N = ⊥ :=
+lemma subsingleton_of_bot : subsingleton (lie_submodule R L ↥(⊥ : lie_submodule R L M)) :=
```
```suggestion
-- TODO[gh-6025]: make this an instance once safe to do so
lemma subsingleton_of_bot : subsingleton (lie_submodule R L ↥(⊥ : lie_submodule R L M)) :=
```
(xref #6025)

eric-wieser commented on src/algebra/lie/basic.lean:
```diff
@@ -1512,22 +1515,19 @@ begin
 In other words, in general, ideals of `I`, regarded as a Lie algebra in its own right, are not the
 same as ideals of `L` contained in `I`. -/
-lemma of_bot_eq_bot (I : lie_ideal R ↥(⊥ : lie_ideal R L)) : I = ⊥ :=
+lemma subsingleton_of_bot : subsingleton (lie_ideal R ↥(⊥ : lie_ideal R L)) :=
```
```suggestion
-- TODO[gh-6025]: make this an instance once safe to do so
lemma subsingleton_of_bot : subsingleton (lie_ideal R ↥(⊥ : lie_ideal R L)) :=
```

eric-wieser commented on src/algebra/lie/basic.lean:
```diff
@@ -2013,6 +2086,13 @@ begin
   exact h₂ hI,
 end
 
+lemma subsingleton_of_semisimple_lie_abelian [is_semisimple R L] [h : is_lie_abelian L] :
```
```suggestion
-- TODO[gh-6025]: make this an instance once safe to do so
lemma subsingleton_of_semisimple_lie_abelian [is_semisimple R L] [h : is_lie_abelian L] :
```

ocfnash committed:
Update src/algebra/lie/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Update src/algebra/lie/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Update src/algebra/lie/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

kim-em commented on src/algebra/lie/basic.lean:
```diff
@@ -553,7 +555,7 @@ variables [comm_ring R] [lie_ring L] [lie_algebra R L] [add_comm_group M] [modul
 
 /-- A Lie module yields a Lie algebra morphism into the linear endomorphisms of the module. -/
-def lie_module.to_endo_morphism : L →ₗ⁅R⁆ module.End R M :=
+@[simps] def lie_module.to_endo_morphism : L →ₗ⁅R⁆ module.End R M :=
```
I know this predates this PR, but could we remove the underscore in the middle of `endo_morphism`?

ocfnash replied:
Done.

For what it's worth, the old name `to_endo_morphism` was supposed to be an abbreviation for `to_endomorphism_morphism` since this is a morphism from the Lie algebra to the endomorphism algebra.

ocfnash committed:
Rename `lie_module.to_endo_morphism` --> `lie_module.to_endomorphism`
